When you're done, compare your work to [these example solutions](https://github.com/Thinkful-Ed/machine-learning-regression-problems/blob/master/notebooks/5.solution_evaluating_goodness_of_fit.ipynb).



## 1. Weather model

For this assignment, you'll revisit the historical temperature dataset.

* First, load the dataset from the **weatherinszeged** table from Thinkful's database.
* Like in the previous checkpoint, build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS.

In [3]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn import linear_model
import seaborn as sns
import statsmodels.api as sm
from sqlalchemy import create_engine

# Display preferences.
%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format

import warnings
warnings.filterwarnings(action="ignore")

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

temperature_df = pd.read_sql_query('select * from weatherinszeged',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

In [6]:
temperature_df['residual_temp'] = temperature_df['temperature'] - temperature_df['apparenttemperature']

# scikit-learn's OLS module
Y = temperature_df.residual_temp
X = temperature_df[['humidity', 'windspeed']]

X = sm.add_constant(X)

sm.OLS(Y,X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          residual_temp   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 26 Jan 2020   Prob (F-statistic):               0.00
Time:                        18:12:39   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.4381      0.021   -115.948      0.000      -2.479      -2.397
humidity       3.0292      0.024    126.479      0.000       2.982       3.076
windspeed      0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why?
+ Both the $R^2$ and adjusted $R^2$ values are **unsatisfactory**. According to this summary, our model is only explaining about 29% of the variance in the $temperature\_df['residual\_temp']$ outcome variable.

> * Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS.

In [7]:
temperature_df['wind_humid'] = temperature_df['humidity'] * temperature_df['windspeed']

X_interact = temperature_df[['humidity', 'windspeed', 'wind_humid']]

X_interact = sm.add_constant(X_interact)

sm.OLS(Y,X_interact).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          residual_temp   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sun, 26 Jan 2020   Prob (F-statistic):               0.00
Time:                        18:22:38   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0839      0.033     -2.511      0.012      -0.149      -0.018
humidity      -0.1775      0.043     -4.133      0.000      -0.262      -0.093
windspeed     -0.0905      0.002    -36.797      0.000      -0.095      -0.086
wind_humid     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Now, what is the R-squared of this model? Does this model improve upon the previous one? 
+ The $R^2$/adjusted $R^2$ is higher than the previous model's, explaining about 34% of the variance in our outcome variable, $temperature\_df['residual\_temp']$.
> * Add *visibility* as an additional explanatory variable to the first model and estimate it. 

In [8]:
# new explanatory variable, 'visibility' added to the original model
X_3 = temperature_df[['humidity', 'windspeed','visibility']]

X_3 = sm.add_constant(X_3)

sm.OLS(Y,X_3).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          residual_temp   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Sun, 26 Jan 2020   Prob (F-statistic):               0.00
Time:                        18:32:54   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.5756      0.028    -56.605      0.000      -1.630      -1.521
humidity       2.6066      0.025    102.784      0.000       2.557       2.656
windspeed      0.1199      0.001    179.014      0.000       0.119       0.121
visibility    -0.0540      0.001    -46.614      0.000      -0.056      -0.052
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.282
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                           0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the *visibility* in terms of the improvement in the adjusted R-squared. Which one is more useful?
+ The adjusted $R^2$ increased form about 29% to a little over 30%. The improvement is not as large as the improvement we witnessed from adding the interactive term which indicates that $'visibility'$ is not as important in explaining information in the target variable.

### Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.
+ According to both he AIC and BIC scores, the model with the interactive term performed the best, reinforcing our previous $R^2$ validation of the model.



##  2. House prices model

In this exercise, you'll work on your house prices model.
* Load the **houseprices** data from Thinkful's database then run your house prices model again.

In [9]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import seaborn as sns
import statsmodels.api as sm

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
houses_df = pd.read_sql_query('select * from houseprices',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

houses_df.head(3)

id  mssubclass mszoning  lotfrontage  lotarea street alley lotshape  \
0   1          60       RL       65.000     8450   Pave  None      Reg   
1   2          20       RL       80.000     9600   Pave  None      Reg   
2   3          60       RL       68.000    11250   Pave  None      IR1   

  landcontour utilities  ... poolarea poolqc fence miscfeature miscval mosold  \
0         Lvl    AllPub  ...        0   None  None        None       0      2   
1         Lvl    AllPub  ...        0   None  None        None       0      5   
2         Lvl    AllPub  ...        0   None  None        None       0      9   

  yrsold  saletype  salecondition  saleprice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  

[3 rows x 81 columns]

In [10]:
# Replace missing numerical values with the mean and missing categoricals ('None') with 'No--' 
# -- = first to letters of feature, ie. NoFe = 'no fence' 
# and the lone missing electrical value with house_df.electrical.mode()

houses_df['electrical'].fillna(str(houses_df['electrical'].mode()), inplace=True)

for column_name in houses_df.columns[3:]:
    for idx, value in enumerate(houses_df[column_name]):
        if pd.isna(value):
            try:
                houses_df[column_name].fillna(houses_df[column_name].mean(), inplace=True)
            except:
                houses_df[column_name].fillna('No' + column_name[:2].capitalize(), inplace=True)

In [11]:
# Creating our dummy variables
columns = ['alley','poolqc','garagetype']

# Create a dataframe with added dummy features
houses_df_dummies = pd.get_dummies(houses_df, columns=columns, drop_first=True)
dummy_columns = [c for c in houses_df_dummies.columns if any([c.startswith(n) for n in columns])]
dummy_columns

['alley_NoAl',
 'alley_Pave',
 'poolqc_Fa',
 'poolqc_Gd',
 'poolqc_NoPo',
 'garagetype_Attchd',
 'garagetype_Basment',
 'garagetype_BuiltIn',
 'garagetype_CarPort',
 'garagetype_Detchd',
 'garagetype_NoGa']

In [12]:
X = houses_df_dummies[dummy_columns]
Y = houses_df_dummies.saleprice

# We need to manually add a constant in statsmodels' sm
X = sm.add_constant(X)

results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.274
Model:                            OLS   Adj. R-squared:                  0.269
Method:                 Least Squares   F-statistic:                     49.80
Date:                Sun, 26 Jan 2020   Prob (F-statistic):           5.54e-93
Time:                        18:57:45   Log-Likelihood:                -18310.
No. Observations:                1460   AIC:                         3.664e+04
Df Residuals:                    1448   BIC:                         3.671e+04
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const               4.617e+05   5.62e+04      8.213      0.000    3.51e+05    5.72e+05
alley_NoAl           1.42e+04   1.01e+04      1.409      0.159   -5569.785     3.4e+04
alley_Pave          3.358e+04   1.43e+04      2.341      0.019    5445.283    6.17e+04
poolqc_Fa          -3.091e+05    6.8e+04     -4.548      0.000   -4.42e+05   -1.76e+05
poolqc_Gd          -3.226e+05    6.2e+04     -5.200      0.000   -4.44e+05   -2.01e+05
poolqc_NoPo        -3.222e+05   4.81e+04     -6.702      0.000   -4.17e+05   -2.28e+05
garagetype_Attchd    4.87e+04   2.79e+04      1.747      0.081   -5972.562    1.03e+05
garagetype_Basment  6920.7635   3.19e+04      0.217      0.828   -5.56e+04    6.94e+04
garagetype_BuiltIn  1.011e+05   2.87e+04      3.522      0.000    4.48e+04    1.57e+05
garagetype_CarPort -4.369e+04   3.58e+04     -1.219      0.223   -1.14e+05    2.66e+04
garagetype_Detchd  -2.049e+04    2.8e+04     -0.733      0.464   -7.53e+04    3.44e+04
garagetype_NoGa    -4.958e+04   2.88e+04     -1.725      0.085   -1.06e+05    6812.576
==============================================================================
Omnibus:                      621.430   Durbin-Watson:                   1.991
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4177.383
Skew:                           1.850   Prob(JB):                         0.00
Kurtosis:                      10.415   Cond. No.                         106.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC. Do you think your model is satisfactory? If so, why?
+ The current model is unsatisfactory with an adjusted $R^2$ of $0.269$ although our F-test p-value indicates that our model is contributing something to the explanation of the outcome variable, and thus, out current features contain some explanatory power regarding the variance of our outcome variable, $saleprice'$. We'll have to compare our AIC and BIC values to subsequent models in order to get a better read on them.

### In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables. For each model you try, get the goodness of fit metrics and compare your models with each other.

In [13]:
houses_df_dummies[dummy_columns].corr()

alley_NoAl  alley_Pave  poolqc_Fa  poolqc_Gd  poolqc_NoPo  \
alley_NoAl               1.000      -0.659      0.010      0.012       -0.018   
alley_Pave              -0.659       1.000     -0.006     -0.008        0.012   
poolqc_Fa                0.010      -0.006      1.000     -0.002       -0.534   
poolqc_Gd                0.012      -0.008     -0.002      1.000       -0.654   
poolqc_NoPo             -0.018       0.012     -0.534     -0.654        1.000   
garagetype_Attchd        0.226      -0.122      0.031      0.037       -0.037   
garagetype_Basment       0.030      -0.020     -0.004     -0.005        0.008   
garagetype_BuiltIn       0.065      -0.043     -0.009     -0.011        0.018   
garagetype_CarPort       0.020      -0.013     -0.003     -0.004        0.005   
garagetype_Detchd       -0.269       0.170     -0.022     -0.027        0.019   
garagetype_NoGa         -0.049      -0.005     -0.009     -0.011        0.017   

                    garagetype_Attchd  garagetype_Basment  garagetype_BuiltIn  \
alley_NoAl                      0.226               0.030               0.065   
alley_Pave                     -0.122              -0.020              -0.043   
poolqc_Fa                       0.031              -0.004              -0.009   
poolqc_Gd                       0.037              -0.005              -0.011   
poolqc_NoPo                    -0.037               0.008               0.018   
garagetype_Attchd               1.000              -0.139              -0.308   
garagetype_Basment             -0.139               1.000              -0.029   
garagetype_BuiltIn             -0.308              -0.029               1.000   
garagetype_CarPort             -0.096              -0.009              -0.020   
garagetype_Detchd              -0.729              -0.069              -0.152   
garagetype_NoGa                -0.294              -0.028              -0.061   

                    garagetype_CarPort  garagetype_Detchd  garagetype_NoGa  
alley_NoAl                       0.020             -0.269           -0.049  
alley_Pave                      -0.013              0.170           -0.005  
poolqc_Fa                       -0.003             -0.022           -0.009  
poolqc_Gd                       -0.004             -0.027           -0.011  
poolqc_NoPo                      0.005              0.019            0.017  
garagetype_Attchd               -0.096             -0.729           -0.294  
garagetype_Basment              -0.009             -0.069           -0.028  
garagetype_BuiltIn              -0.020             -0.152           -0.061  
garagetype_CarPort               1.000             -0.047           -0.019  
garagetype_Detchd               -0.047              1.000           -0.146  
garagetype_NoGa                 -0.019             -0.146            1.000

In [38]:
# removing the features with non-statistically significant p-values and adding two continuous variables, 
# 'lotarea' and 'totalbsmtsf' to our regressors/explanatory variables
new_variables = ['alley_Pave','poolqc_Fa','poolqc_Gd','poolqc_NoPo','garagetype_Attchd','garagetype_BuiltIn',
                 'garagetype_NoGa','lotarea','totalbsmtsf']

# Checking for correlation
houses_df_dummies[new_variables].corr()

alley_Pave  poolqc_Fa  poolqc_Gd  poolqc_NoPo  \
alley_Pave               1.000     -0.006     -0.008        0.012   
poolqc_Fa               -0.006      1.000     -0.002       -0.534   
poolqc_Gd               -0.008     -0.002      1.000       -0.654   
poolqc_NoPo              0.012     -0.534     -0.654        1.000   
garagetype_Attchd       -0.122      0.031      0.037       -0.037   
garagetype_BuiltIn      -0.043     -0.009     -0.011        0.018   
garagetype_NoGa         -0.005     -0.009     -0.011        0.017   
lotarea                 -0.078      0.011      0.094       -0.087   
totalbsmtsf             -0.056      0.013      0.163       -0.151   

                    garagetype_Attchd  garagetype_BuiltIn  garagetype_NoGa  \
alley_Pave                     -0.122              -0.043           -0.005   
poolqc_Fa                       0.031              -0.009           -0.009   
poolqc_Gd                       0.037              -0.011           -0.011   
poolqc_NoPo                    -0.037               0.018            0.017   
garagetype_Attchd               1.000              -0.308           -0.294   
garagetype_BuiltIn             -0.308               1.000           -0.061   
garagetype_NoGa                -0.294              -0.061            1.000   
lotarea                         0.123               0.038           -0.074   
totalbsmtsf                     0.381              -0.032           -0.160   

                    lotarea  totalbsmtsf  
alley_Pave           -0.078       -0.056  
poolqc_Fa             0.011        0.013  
poolqc_Gd             0.094        0.163  
poolqc_NoPo          -0.087       -0.151  
garagetype_Attchd     0.123        0.381  
garagetype_BuiltIn    0.038       -0.032  
garagetype_NoGa      -0.074       -0.160  
lotarea               1.000        0.261  
totalbsmtsf           0.261        1.000

In [40]:
X_new = houses_df_dummies[new_variables]

X_new = sm.add_constant(X_new)

sm.OLS(Y,X_new).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.516
Model:                            OLS   Adj. R-squared:                  0.513
Method:                 Least Squares   F-statistic:                     171.7
Date:                Sun, 26 Jan 2020   Prob (F-statistic):          3.43e-221
Time:                        19:42:26   Log-Likelihood:                -18014.
No. Observations:                1460   AIC:                         3.605e+04
Df Residuals:                    1450   BIC:                         3.610e+04
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const               2.818e+05   3.98e+04      7.087      0.000    2.04e+05     3.6e+05
alley_Pave          2.305e+04   8913.165      2.587      0.010    5569.905    4.05e+04
poolqc_Fa          -2.247e+05   5.56e+04     -4.045      0.000   -3.34e+05   -1.16e+05
poolqc_Gd           -3.84e+05   5.07e+04     -7.575      0.000   -4.83e+05   -2.85e+05
poolqc_NoPo         -2.34e+05   3.94e+04     -5.942      0.000   -3.11e+05   -1.57e+05
garagetype_Attchd   3.478e+04   3573.197      9.733      0.000    2.78e+04    4.18e+04
garagetype_BuiltIn  1.043e+05   6564.351     15.892      0.000    9.14e+04    1.17e+05
garagetype_NoGa    -2.238e+04   6746.181     -3.318      0.001   -3.56e+04   -9150.616
lotarea                0.7483      0.152      4.938      0.000       0.451       1.046
totalbsmtsf           93.4267      3.744     24.953      0.000      86.082     100.771
==============================================================================
Omnibus:                      414.863   Durbin-Watson:                   1.928
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3779.778
Skew:                           1.054   Prob(JB):                         0.00
Kurtosis:                      10.595   Cond. No.                     8.31e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.31e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [57]:
# creating another model with a random variable column
houses_df_dummies['random'] = np.random.rand(len(houses_df_dummies),1)

X_rand = houses_df_dummies[new_variables + ['random']]
X_rand = sm.add_constant(X_rand)

sm.OLS(Y,X_rand).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.517
Model:                            OLS   Adj. R-squared:                  0.514
Method:                 Least Squares   F-statistic:                     155.1
Date:                Sun, 26 Jan 2020   Prob (F-statistic):          8.49e-221
Time:                        19:55:02   Log-Likelihood:                -18013.
No. Observations:                1460   AIC:                         3.605e+04
Df Residuals:                    1449   BIC:                         3.611e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const               2.883e+05   3.99e+04      7.228      0.000     2.1e+05    3.67e+05
alley_Pave          2.302e+04   8905.782      2.585      0.010    5547.799    4.05e+04
poolqc_Fa          -2.271e+05   5.55e+04     -4.089      0.000   -3.36e+05   -1.18e+05
poolqc_Gd          -3.857e+05   5.07e+04     -7.613      0.000   -4.85e+05   -2.86e+05
poolqc_NoPo         -2.36e+05   3.94e+04     -5.997      0.000   -3.13e+05   -1.59e+05
garagetype_Attchd   3.456e+04   3572.253      9.674      0.000    2.76e+04    4.16e+04
garagetype_BuiltIn  1.046e+05   6560.292     15.940      0.000    9.17e+04    1.17e+05
garagetype_NoGa    -2.257e+04   6741.329     -3.348      0.001   -3.58e+04   -9346.211
lotarea                0.7574      0.151      4.999      0.000       0.460       1.055
totalbsmtsf           93.5558      3.742     25.004      0.000      86.216     100.895
random             -9262.8818   5014.373     -1.847      0.065   -1.91e+04     573.326
==============================================================================
Omnibus:                      409.407   Durbin-Watson:                   1.928
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3647.318
Skew:                           1.044   Prob(JB):                         0.00
Kurtosis:                      10.456   Cond. No.                     8.31e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.31e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Which model is the best and why?

+ Disappointingly, our model with the random feature outperformed our $'new\_variables'$ model in both $R^2$, which was expected due to the larger amount of features, but also the adjusted $R^2$, which should have been less affected by it, albeit by only a slight margin. Both of them outperformed our original model with an adjusted $R^2$ of almost 52% compared to just 27%. 

+ That being said, the BIC score is slighly lower without the random feature. Thus, our $'new\_variables'$ model without the random feature, performed the best. Notably, our random feature is also the only feature that does not have a p-value below $0.05$.